### PR TITLE
Switch the Windows installer from NSIS to MSI

### PR DESCRIPTION
**Pull request by hughsie**
This means we get a few bugs fixed:

 * We appear in the Add/Remove Programs UI
 * We have an uninstaller that works on Windows 10
 * The installation is pretty, and scriptable
 * We actually install the translations

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:31]: I'd actually argue we don't want bash completion either on Windows

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:59]: do any other subprojects need tests turned off too for this?

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:144]: I think from this point onwards it would be good to only run the code if `CI=1`.  That way we don't need to make the MSI for "local" builds.

File: meson.build
Comment (superm1) [inline meson.build:527]: <cough> arm </cough>

File: meson.build
Comment (superm1) [inline meson.build:568]: I think you want to add a `find_program('wixl', required: host_machine.system == 'windows')` here still

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:118]: although I really like how programattic this list is, this change breaks running without installing.  As it stands today you can drop a `dist` directory onto a real Windows host (or "natively" launch it from WSL2) and it runs.  I want to maintain that if at all possible.

Can we maybe make symlinks for all the files and just not install the symlinks from the MSI?

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:61]: any idea why these are still made?  or that's TODO?

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:59]: I think we turn them off for the "main" configure step already.

File: meson.build
Comment (hughsie) [reply meson.build:527]: I'm guessing it's `host_machine.cpu() of "aarch64" to "arm64"`?

File: meson.build
Comment (hughsie) [reply meson.build:568]: I figured it was just a non-installed configure.in-style replacement -- your logic is just to avoid doing that on win32 if it's not going to be used?

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:118]: would copying those files into `$DESTDIR/bin` also work? That does feel a bit of a hack. Can you describe a bit how you use WSL right now so I can make sure this doesn't break things?

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:61]: deadcode, removed.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:31]: Agree, fixed.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:144]: I'm not sure much of contrib/ci/build_windows.sh makes sense outside of CI -- it unconditionally `rm -rf`'s your build-win32 for instance. Maybe we make the script quit right at the start if CI != 1?

Comment (superm1) [reply contrib/ci/build_windows.sh:151]: If we want people to be able to make changes to Win32, we have to make building for Win32 hosts accessible though..

File: meson.build
Comment (superm1) [reply meson.build:568]: My thought is that if the build script needs it, then it should be something that we check for in meson.build so we can message accordingly.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:151]: I don't disagree, but I think that's a different script to the CI one.

File: meson.build
Comment (hughsie) [reply meson.build:568]: Well, the CI script needs it, but not generally for building win32...

File: contrib/ci/build_windows.sh
Comment (superm1) [reply contrib/ci/build_windows.sh:125]: What i've been doing is:
1. Open up Fedora in WSL
2. `./contrib/ci/build_windows.sh`
3. `./dist/bin/fwupdtool.exe`

I don't mind changing that workflow a bit (add environment variable here, or there etc), but adding a step of installing an MSI is too much to me.

File: .circleci/config.yml
Comment (superm1) [inline .circleci/config.yml:37]: drop a comment on why this specific version is needed so we know when to switch away?

File: contrib/ci/build_windows.sh
Comment (superm1) [reply contrib/ci/build_windows.sh:125]: Copying those files to `$DESTDIR/bin` would also work.  Or I wonder if maybe there is an easy way to just extract the MSI and run it from that directory it's extracted to?

File: contrib/ci/build_windows.sh
Comment (superm1) [reply contrib/ci/build_windows.sh:151]: We do have other scripts that change one or two things based on CI=1 though.

File: .circleci/config.yml
Comment (hughsie) [reply .circleci/config.yml:37]: I've asked the msitools maintainer to do a new upstream release, so maybe we can just wait until the official release happens.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh:125]: Or is it possible on Windows to set the DLL search path add /usr/x86_64-w64-mingw32/sys-root/mingw/bin`?

File: .circleci/config.yml
Comment (superm1) [inline .circleci/config.yml:57]: AFAICT this MSI is ending up in `./setup` not `dist/setup`

File: contrib/ci/build_windows.sh
Comment (superm1) [inline contrib/ci/build_windows.sh:150]: needs `DESTDIR` prefix

Comment (hughsie) [reply contrib/ci/build_windows.sh:150]: Ack fixed, thanks.

File: contrib/ci/build_windows.sh
Comment (superm1) [reply contrib/ci/build_windows.sh:125]: Not easily that I can tell.   How about if you just run `msiextract` at the end of the script?  It should overwrite everything in dist and copy in the missing dependency libraries.

Comment (hughsie) [reply contrib/ci/build_windows.sh]: That didn't work super-well. I've added a more useful script in ./contrib -- can you try that with WSL and Fedora please.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh]: I think it actually makes sense to have different compile options for the CI and a local win32 build. I've added a script for local "getting up and going" use.

File: contrib/build-windows.sh
Comment (superm1) [inline contrib/build-windows.sh:36]: How about just source them from one another rather than have to keep in sync?

**Comment by cwsystem76** [comment]
Via the [artifact build](https://app.circleci.com/pipelines/github/fwupd/fwupd/9892/workflows/11961248-0d10-4686-be25-d3822d24dd64/jobs/35847/artifacts) in the CI:

image

And the uninstaller works as intended from the "Apps & features" page and from the old Control Panel as well. One thing to note, though, it's being installed into `Program Files (x86)`, which is for 32-bit apps, instead of `Program Files` which is where 64bit apps usually go. There may be a reason for this but I thought I'd mention it.

File: contrib/build-windows.sh
Comment (hughsie) [reply contrib/build-windows.sh:25]: I did ponder putting them in `contrib/ci/dependencies.xml` -- e.g.

      <package variant="mingw64">mingw64-curl</package>

insane?

**Comment by hughsie** [comment]
> it's being installed into Program Files (x86)

It seems that although we specify `Win64=true` in the `Component` and set `InstalledVersion=200` we're hitting https://gitlab.gnome.org/GNOME/msitools/-/issues/46

File: contrib/build-windows.sh
Comment (superm1) [reply contrib/build-windows.sh:25]: Not insane, makes perfect sense!

Comment (hughsie) [reply contrib/build-windows.sh:14]: Ya,concur, it kinda works.

Comment (superm1) [reply contrib/build-windows.sh:14]: Drop the manual install steps now from circleci too.

Comment (superm1) [reply contrib/build-windows.sh:14]: Drop the manual install steps now from circleci too.

Comment (superm1) [reply contrib/build-windows.sh:14]: On my phone maybe i missed it. I was meaning  in .circleci/config.yaml

File: .circleci/config.yml
Comment (superm1) [inline .circleci/config.yml:37]: shouldn't everything else above here be in dependencies.xml?  The idea is supposed to be you don't have to manually install anything

Comment (hughsie) [reply .circleci/config.yml]: Yes, and done. There are a few more native things than I liked, but I've been filing bugs against other components all day. As they get merged and we update the subprojects we can remove a few BRs from the mingw64 build.

File: .circleci/config.yml
Comment (hughsie) [reply .circleci/config.yml]: I'll rip out the manual dnf install https:// when the update hits updates-testing in a few days.

File: contrib/ci/build_windows.sh
Comment (hughsie) [reply contrib/ci/build_windows.sh]: I actually got WSL working with Fedora too. :)

File: contrib/build-windows.sh
Comment (superm1) [inline contrib/build-windows.sh:26]: what do you think about converting these two to tristate options that auto-disable for os == windows?

Comment (hughsie) [reply contrib/build-windows.sh:26]: do either *ever* make sense on Windows? If not, then yes please, although I think you're the expert at the auto-disable magic.

**Review by superm1 — APPROVED** [review]
I still do think running ./contrib/build-windows.sh from contrib/ci/build-windows.sh would make sense, but otherwise it LGTM.

File: contrib/build-windows.sh
Comment (superm1) [reply contrib/build-windows.sh:26]: The only time you run bash in Windows is WSL, and if you're in WSL you won't read bash-completion.  So I don't think they ever make sense in Windows

**Comment by hughsie** [comment]
> still do think running ./contrib/build-windows.sh from contrib/ci/build-windows.sh would make sense, 

It's not a 1:1 mapping; i.e. some options make sense in CI, but not for local hacking and vice-versa.

File: contrib/ci/build_windows.sh
Comment (superm1) [reply contrib/ci/build_windows.sh]: It's handy right!

**Pull request merged**
